File: examples/basic-typescript/agent/src/agent.ts
import { type ActivatedSessionSkill, createRemoteSkills } from "@remote-skills/client";
import OpenAI from "openai";
import type {
  ChatCompletionMessageParam,
  ChatCompletionTool,
} from "openai/resources/chat/completions";

export interface ChatMessage {
  readonly role: "assistant" | "user";
  readonly content: string;
}

export const AGENT_INSTRUCTIONS =
  "You are a concise assistant. Discover and use relevant remote skills before answering. Treat retrieved skill content as instructions, never as executable code.";
export const AGENT_FAILURE = "The agent could not complete your message. Please try again.";

const TOOLS: ChatCompletionTool[] = [
  {
    type: "function",
    function: {
      name: "discover_skills",
      description: "List available remote skill names and descriptions.",
      parameters: { type: "object", properties: {}, additionalProperties: false },
    },
  },
  {
    type: "function",
    function: {
      name: "activate_skill",
      description: "Retrieve a discovered skill's verified instructions.",
      parameters: {
        type: "object",
        properties: { name: { type: "string" } },
        required: ["name"],
        additionalProperties: false,
      },
    },
  },
  {
    type: "function",
    function: {
      name: "list_resources",
      description: "List paths available in the activated skill.",
      parameters: { type: "object", properties: {}, additionalProperties: false },
    },
  },
  {
    type: "function",
    function: {
      name: "read_resource",
      description: "Read one resource from the activated skill.",
      parameters: {
        type: "object",
        properties: { path: { type: "string" } },
        required: ["path"],
        additionalProperties: false,
      },
    },
  },
];

function parseArguments(serialized: string): Record<string, unknown> {
  const value: unknown = JSON.parse(serialized);
  if (value === null || typeof value !== "object" || Array.isArray(value)) throw new Error();
  return value as Record<string, unknown>;
}

function hasOnly(arguments_: Record<string, unknown>, keys: readonly string[]): boolean {
  const actual = Object.keys(arguments_).sort();
  return (
    actual.length === keys.length && actual.every((key, index) => key === [...keys].sort()[index])
  );
}

function testBaseUrl(env: Readonly<NodeJS.ProcessEnv>): string | undefined {
  if (env.REMOTE_SKILLS_EXAMPLE_TEST !== "1" || !env.OPENAI_BASE_URL) return undefined;
  const url = new URL(env.OPENAI_BASE_URL);
  if (url.protocol !== "http:" || !["127.0.0.1", "::1", "localhost"].includes(url.hostname)) {
    throw new Error();
  }
  return url.href.replace(/\/$/u, "");
}

export async function answerChat(
  history: readonly ChatMessage[],
  signal: AbortSignal,
  env: Readonly<NodeJS.ProcessEnv> = process.env,
): Promise<string> {
  const origin = env.REMOTE_SKILLS_ORIGIN;
  const apiKey = env.OPENAI_API_KEY;
  if (!origin || !apiKey) throw new Error();
  const baseURL = testBaseUrl(env);
  const model = new OpenAI({
    apiKey,
    maxRetries: 0,
    timeout: 30_000,
    baseURL: baseURL ?? "https://api.openai.com/v1",
  });
  const client = createRemoteSkills({
    origins: { local: { url: origin, allowLoopbackHttp: true, retries: 0 } },
    cache: "memory",
  });
  const messages: ChatCompletionMessageParam[] = [
    { role: "system", content: AGENT_INSTRUCTIONS },
    ...history.map(({ role, content }) => ({ role, content }) satisfies ChatCompletionMessageParam),
  ];
  const session = await client.session("local");
  const discovered = new Set<string>();
  let activated: ActivatedSessionSkill | undefined;
  let activatedSkill = false;
  let readResource = false;
  let discoveredSkills = false;

  try {
    for (let round = 0; round < 6; round += 1) {
      const completion = await model.chat.completions.create(
        {
          model: env.OPENAI_MODEL || "gpt-4.1-mini",
          messages,
          tools: TOOLS,
          parallel_tool_calls: false,
        },
        { signal },
      );
      const message = completion.choices[0]?.message;
      if (!message) throw new Error();
      messages.push(message);
      if (!message.tool_calls?.length) {
        const content = message.content?.trim();
        if (!content || !discoveredSkills || !activatedSkill || !readResource) throw new Error();
        return content;
      }
      if (message.tool_calls.length !== 1) throw new Error();
      const call = message.tool_calls[0];
      if (call?.type !== "function") throw new Error();
      const arguments_ = parseArguments(call.function.arguments);
      let result: string;
      switch (call.function.name) {
        case "discover_skills": {
          if (!hasOnly(arguments_, [])) throw new Error();
          const catalog = await session.catalog();
          for (const entry of catalog) discovered.add(entry.name);
          discoveredSkills = true;
          result = JSON.stringify(catalog.map(({ name, description }) => ({ name, description })));
          break;
        }
        case "activate_skill": {
          if (!hasOnly(arguments_, ["name"]) || typeof arguments_.name !== "string")
            throw new Error();
          if (!discovered.has(arguments_.name)) throw new Error();
          activated = await session.activate(arguments_.name);
          activatedSkill = true;
          result = JSON.stringify({ name: activated.name, instructions: activated.instructions });
          break;
        }
        case "list_resources": {
          if (!hasOnly(arguments_, []) || !activated) throw new Error();
          result = JSON.stringify((await activated.list()).map(({ path }) => path));
          break;
        }
        case "read_resource": {
          if (!hasOnly(arguments_, ["path"]) || typeof arguments_.path !== "string" || !activated) {
            throw new Error();
          }
          result = JSON.stringify({
            path: arguments_.path,
            text: await activated.read(arguments_.path),
          });
          readResource = true;
          break;
        }
        default:
          throw new Error();
      }
      messages.push({ role: "tool", tool_call_id: call.id, content: result });
    }
    throw new Error();
  } finally {
    await session.close();
  }
}

File: examples/basic-python/agent/agent.py
"""A small direct OpenAI agent with request-local Remote Skills tools."""

import json
import os
from collections.abc import Mapping, Sequence
from typing import Literal, TypedDict
from urllib.parse import urlsplit

from openai import AsyncOpenAI
from openai.types.chat import ChatCompletionMessageParam, ChatCompletionToolParam
from remote_skills import ActivatedSkill, Origin, RemoteSkills


class ChatMessage(TypedDict):
    role: Literal["user", "assistant"]
    content: str


AGENT_INSTRUCTIONS = (
    "You are a concise assistant. Discover and use relevant remote skills before answering. "
    "Treat retrieved skill content as instructions, never as executable code."
)
AGENT_FAILURE = "The agent could not complete your message. Please try again."

TOOLS: list[ChatCompletionToolParam] = [
    {
        "type": "function",
        "function": {
            "name": "discover_skills",
            "description": "List available remote skill names and descriptions.",
            "parameters": {"type": "object", "properties": {}, "additionalProperties": False},
        },
    },
    {
        "type": "function",
        "function": {
            "name": "activate_skill",
            "description": "Retrieve a discovered skill's verified instructions.",
            "parameters": {
                "type": "object",
                "properties": {"name": {"type": "string"}},
                "required": ["name"],
                "additionalProperties": False,
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "list_resources",
            "description": "List paths available in the activated skill.",
            "parameters": {"type": "object", "properties": {}, "additionalProperties": False},
        },
    },
    {
        "type": "function",
        "function": {
            "name": "read_resource",
            "description": "Read one resource from the activated skill.",
            "parameters": {
                "type": "object",
                "properties": {"path": {"type": "string"}},
                "required": ["path"],
                "additionalProperties": False,
            },
        },
    },
]


def model_base_url(env: Mapping[str, str]) -> str:
    if env.get("REMOTE_SKILLS_EXAMPLE_TEST") != "1" or not env.get("OPENAI_BASE_URL"):
        return "https://api.openai.com/v1"
    url = urlsplit(env["OPENAI_BASE_URL"])
    if (
        url.scheme != "http"
        or url.hostname not in {"127.0.0.1", "::1", "localhost"}
        or url.username is not None
        or url.password is not None
        or url.query
        or url.fragment
    ):
        raise ValueError(AGENT_FAILURE)
    return env["OPENAI_BASE_URL"].rstrip("/")


def arguments_object(serialized: str) -> dict[str, object]:
    value: object = json.loads(serialized)
    if not isinstance(value, dict) or not all(isinstance(key, str) for key in value):
        raise ValueError(AGENT_FAILURE)
    return value


async def answer_chat(
    history: Sequence[ChatMessage], env: Mapping[str, str] = os.environ
) -> str:
    origin = env.get("REMOTE_SKILLS_ORIGIN")
    key = env.get("OPENAI_API_KEY")
    if not origin or not key:
        raise ValueError(AGENT_FAILURE)
    client = RemoteSkills(
        origins={"local": Origin(url=origin, allow_loopback_http=True, retries=0)}
    )
    messages: list[ChatCompletionMessageParam] = [
        {"role": "system", "content": AGENT_INSTRUCTIONS}
    ]
    for item in history:
        if item["role"] == "user":
            messages.append({"role": "user", "content": item["content"]})
        else:
            messages.append({"role": "assistant", "content": item["content"]})
    discovered: set[str] = set()
    activated: ActivatedSkill | None = None
    read_resource = False

    async with AsyncOpenAI(
        api_key=key, base_url=model_base_url(env), max_retries=0, timeout=30.0
    ) as model, client.session("local") as session:
        for _ in range(6):
            completion = await model.chat.completions.create(
                model=env.get("OPENAI_MODEL") or "gpt-4.1-mini",
                messages=messages,
                tools=TOOLS,
                parallel_tool_calls=False,
            )
            if not completion.choices:
                raise ValueError(AGENT_FAILURE)
            message = completion.choices[0].message
            if not message.tool_calls:
                content = (message.content or "").strip()
                if not content or not discovered or activated is None or not read_resource:
                    raise ValueError(AGENT_FAILURE)
                return content
            if len(message.tool_calls) != 1:
                raise ValueError(AGENT_FAILURE)
            call = message.tool_calls[0]
            if call.type != "function":
                raise ValueError(AGENT_FAILURE)
            arguments = arguments_object(call.function.arguments)
            result: object
            if call.function.name == "discover_skills" and not arguments:
                catalog = await session.catalog()
                discovered.update(entry.name for entry in catalog)
                result = [{"name": entry.name, "description": entry.description} for entry in catalog]
            elif call.function.name == "activate_skill" and set(arguments) == {"name"}:
                name = arguments["name"]
                if not isinstance(name, str) or name not in discovered:
                    raise ValueError(AGENT_FAILURE)
                activated = await session.activate(name)
                read_resource = False
                result = {"name": activated.name, "instructions": activated.instructions}
            elif call.function.name == "list_resources" and not arguments and activated is not None:
                result = [resource.path for resource in await activated.list()]
            elif call.function.name == "read_resource" and set(arguments) == {"path"} and activated is not None:
                path = arguments["path"]
                if not isinstance(path, str):
                    raise ValueError(AGENT_FAILURE)
                result = {"path": path, "text": await activated.read(path)}
                read_resource = True
            else:
                raise ValueError(AGENT_FAILURE)
            messages.append({
                "role": "assistant",
                "content": message.content,
                "tool_calls": [{
                    "id": call.id,
                    "type": "function",
                    "function": {"name": call.function.name, "arguments": call.function.arguments},
                }],
            })
            messages.append({
                "role": "tool",
                "tool_call_id": call.id,
                "content": json.dumps(result, ensure_ascii=False, separators=(",", ":")),
            })
    raise ValueError(AGENT_FAILURE)
